File: 03/sol.py
def read_file(filename="input.txt"):
    with open("input.txt", "r") as f:
        return f.read().strip()


def parse(input_string):
    lines = input_string.split("\n")
    return lines


def intersections(bags):
    return set.intersection(*(set(b) for b in bags)).pop()


def points(char):
    assert len(char) == 1
    if char.islower():
        return ord(char) - 96
    if char.isupper():
        return ord(char) - 64 + 26
    return 0


def solve1(lines):
    bags = [(line[:len(line)//2], line[len(line)//2:]) for line in lines]
    chars = [intersections(bag) for bag in bags]
    return sum(points(char) for char in chars)


def solve2(lines):
    bags = [lines[(i*3):(i*3+3)] for i in range(len(lines)//3)]
    chars = [intersections(bag) for bag in bags]
    return sum(points(line) for line in chars)


if __name__ == "__main__":
    input_string = read_file("input.txt")
    assert input_string is not None
    lines = parse(input_string)
    res1 = solve1(lines)
    print(res1)
    res2 = solve2(lines)
    print(res2)

File: 03/sol.jl
function parseinput1(input)
    lines = split(input, "\n")
    sets = map(line -> (Set(line[begin:end÷2]), Set(line[end÷2+1:end])), lines)
    # sets = [(Set(line[begin:end÷2]), Set(line[end÷2+1:end])) for line in lines]
    return sets
end

function parseinput2(input)
    lines = split(input, '\n')
    sets = map(Set, lines)
    mat = reshape(sets, 3, :)
    return mat
end

function points(x::Char)::Int
    if islowercase(x)
        return x - 'a' + 1
    end
    if isuppercase(x)
        return x - 'A' + 27
    end
end

function solve1(x)
    return sum(points(c) for c in map(t -> pop!(t[1] ∩ t[2]), x))
end

function solve2(x)
    sum(points(c) for c in mapslices(v -> pop!(∩(v...)), x, dims=1))
end

if abspath(PROGRAM_FILE) == @__FILE__
    filename = isempty(ARGS) ? "input.txt" : ARGS[1]
    input = rstrip(read(filename, String))
    x1 = parseinput1(input)
    r1 = solve1(x1)
    println(r1)
    x2 = parseinput2(input)
    r2 = solve2(x2)
    println(r2)
end
